journal: enforce strict consistency for realtime timestamps on write

msizanoen1 opened the pull request:
Ensure all realtime timestamps in a journal file are strictly ordered on write as a defense-in-depth measure. All known callers of journal_file_append_entry and journal_file_copy_entry, which call this function, should be able to handle the error by rotating the journal.

This is especially helpful for systems with RTC local time enabled, where all log entries from initramfs might be recorded as several hours later than it actually is, which won't get caught by journald during log flushing. In those cases, the resulting inconsistency can cause libsystemd to loop infinitely through journal files as observed in `abrt-dump-journal-oops`.

This also fix a potential issue with bisecting journal files created by `systemd-journal-remote`.

## How to reproduce
 - Set local timezone to a positive timezone
 - `timedatectl set-local-rtc 1`
 - Reboot
 - `for file in /var/log/journal/*/*; do journalctl -o json-seq --file $file | jq -r --seq '.__REALTIME_TIMESTAMP' | sort -cn; done`

## Expected results
The last command shouldn't output anything

## Actual results
The last command results in a `sort` disorder error

Prerequisite: #24996

yuwata commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -2026,6 +2026,12 @@ static int journal_file_append_entry_internal(
+                return log_debug_errno(SYNTHETIC_ERRNO(ETXTBSY),
+                                       "Realtime timestamp %" PRIu64 " smaller than previous realtime "
+                                       "timestamp %" PRIu64 ", refusing entry.",
+                                       ts->realtime, le64toh(f->header->tail_entry_realtime));
```
Hmm,,, it sounds that this introduces too many log rotations when an NTP client is running. No?

msizanoen1 replied:
The normal write path is already covered by this: https://github.com/systemd/systemd/blob/f4a4493eb57bffec091c8f0d4bb846850e544aa4/src/journal/journald-server.c#L837-L845

This change is mostly intended as a defense-in-depth measure and extend it to cover runtime-to-persistent log flushing at boot time.

msizanoen1 replied:
This is unrelated to the CI failure caused by `systemd-journal-remote`, which is fixed in #24996. Behavior for `systemd-journald` should remain unchanged.

mrc0mmand commented:
I'm not sure what's going on, but this seems to generate a metric ton of journal files:

```
# ls -lh
total 494M
-rw-r--r--.  1 root root 500M Oct 11 18:46 basic.img
drwxr-xr-x. 16 root root 1.0K Oct 11 18:42 root
drwxr-xr-x. 14 root root 4.0K Oct 11 18:40 unprivileged-nspawn-root
# ls -lh root/var/log/journal/f6737358e3cc4c308f112364385fb8cd/
total 10M
-rw-r-----+ 1 root systemd-journal 2.1M Oct 11 18:42 system@fff01d9e0d854cc7907c6440cc2837e8-0000000000000001-0005eac69ecd9367.journal
-rw-r-----+ 1 root systemd-journal 2.6M Oct 11 18:42 system@fff01d9e0d854cc7907c6440cc2837e8-0000000000000e47-0005eac6a1ac30c0.journal
-rw-r-----+ 1 root systemd-journal 5.4M Oct 11 18:42 system.journal
# ~/systemd/build/systemd-journal-remote -o test.journal --getter "/root/systemd/build/journalctl -o export -D root/var/log/journal"
Spawning getter /root/systemd/build/journalctl -o export -D root/var/log/journal...
Finishing after writing 7333 entries
# ls -l
total 934561
-rw-r--r--.  1 root root 524288000 Oct 11 18:46 basic.img
drwxr-xr-x. 16 root root      1024 Oct 11 18:42 root
-rw-r-----.  1 root root   3942104 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000001-0005eac69ecd9367.journal
-rw-r-----.  1 root root   3740744 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000001cf-0005eac69ece5db0.journal
-rw-r-----.  1 root root   3741600 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000001d2-0005eac69ecea7e8.journal
-rw-r-----.  1 root root   3756432 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000001d4-0005eac69ecdcf13.journal
-rw-r-----.  1 root root   3751680 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000001eb-0005eac69efc8a00.journal
-rw-r-----.  1 root root   3740960 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000001f8-0005eac69f28c250.journal
-rw-r-----.  1 root root   3753904 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000001fc-0005eac69f2904b8.journal
-rw-r-----.  1 root root   3791072 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000208-0005eac69f4648e8.journal
-rw-r-----.  1 root root   3745608 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000024c-0005eac69f8fa628.journal
-rw-r-----.  1 root root   3753392 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000251-0005eac69f95e7b8.journal
-rw-r-----.  1 root root   3751128 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000025d-0005eac69fa3fd30.journal
-rw-r-----.  1 root root   3766096 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000268-0005eac69fb9ea78.journal
-rw-r-----.  1 root root   3763816 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000284-0005eac69fe78258.journal
-rw-r-----.  1 root root   3746496 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000029a-0005eac69fe8e9b8.journal
-rw-r-----.  1 root root   3752216 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000029f-0005eac69fe91898.journal
-rw-r-----.  1 root root   3744952 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000002ab-0005eac69fea6479.journal
-rw-r-----.  1 root root   3745232 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000002af-0005eac69fea8729.journal
-rw-r-----.  1 root root   3743616 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000002b3-0005eac69feaaaf0.journal
-rw-r-----.  1 root root   3769088 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000002b6-0005eac69feab20b.journal
-rw-r-----.  1 root root   3981432 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000002d6-0005eac69fedeaf8.journal
-rw-r-----.  1 root root   3742264 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004a2-0005eac6a00d4e20.journal
-rw-r-----.  1 root root   3746304 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004ac-0005eac6a039d878.journal
-rw-r-----.  1 root root   3746896 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004b6-0005eac6a0430fd8.journal
-rw-r-----.  1 root root   3750456 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004c1-0005eac6a0debf50.journal
-rw-r-----.  1 root root   3742784 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004ca-0005eac6a0e78568.journal
-rw-r-----.  1 root root   3739800 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004cd-0005eac6a034e780.journal
-rw-r-----.  1 root root   3739752 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004cf-0005eac6a034f042.journal
-rw-r-----.  1 root root   3741104 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004d1-0005eac6a0350907.journal
-rw-r-----.  1 root root   3747280 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004d4-0005eac6a03512c6.journal
-rw-r-----.  1 root root   3793784 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000004e0-0005eac6a036b9ee.journal
-rw-r-----.  1 root root   3756552 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000053b-0005eac6a036edd0.journal
-rw-r-----.  1 root root   3753384 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000556-0005eac6a036f12b.journal
-rw-r-----.  1 root root   3771504 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000568-0005eac6a036f932.journal
-rw-r-----.  1 root root   3779464 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000592-0005eac6a0370eb2.journal
-rw-r-----.  1 root root   3901184 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000005cd-0005eac6a0372824.journal
-rw-r-----.  1 root root   3930176 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000006bf-0005eac6a037dbe0.journal
-rw-r-----.  1 root root   3760232 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000007e6-0005eac6a1062a65.journal
-rw-r-----.  1 root root   3761808 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000007fa-0005eac6a106898c.journal
-rw-r-----.  1 root root   3835552 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000819-0005eac6a1074983.journal
-rw-r-----.  1 root root   3763200 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000088f-0005eac6a10ff8dc.journal
-rw-r-----.  1 root root   3791104 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000008b0-0005eac6a111d671.journal
-rw-r-----.  1 root root   3746176 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000008ef-0005eac6a1135232.journal
-rw-r-----.  1 root root   3954176 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000008f4-0005eac6a11374a6.journal
-rw-r-----.  1 root root   3741384 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000a57-0005eac6a12c8578.journal
-rw-r-----.  1 root root   3751664 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000a58-0005eac6a12c7cfc.journal
-rw-r-----.  1 root root   3751128 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000a63-0005eac6a1881dee.journal
-rw-r-----.  1 root root   3741000 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000a70-0005eac6a18870b8.journal
-rw-r-----.  1 root root   3790704 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000a73-0005eac6a188438b.journal
-rw-r-----.  1 root root   3740824 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000ab0-0005eac6a18b8d98.journal
-rw-r-----.  1 root root   3753120 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000ab1-0005eac6a18b74f4.journal
-rw-r-----.  1 root root   3765312 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000abd-0005eac6a1920ed3.journal
-rw-r-----.  1 root root   3738176 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000adc-0005eac6a1931360.journal
-rw-r-----.  1 root root   3740088 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000adf-0005eac6a19326e8.journal
-rw-r-----.  1 root root   3753168 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000ae1-0005eac6a192b558.journal
-rw-r-----.  1 root root   3841896 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000af6-0005eac6a192d421.journal
-rw-r-----.  1 root root   3756520 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000b7c-0005eac6a1947a06.journal
-rw-r-----.  1 root root   3792688 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000b96-0005eac6a19483fe.journal
-rw-r-----.  1 root root   4122272 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000bed-0005eac6a194b2dc.journal
-rw-r-----.  1 root root   3742304 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000e4a-0005eac6a1b05b78.journal
-rw-r-----.  1 root root   3758288 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000e4c-0005eac6a1a14eec.journal
-rw-r-----.  1 root root   3760448 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000e63-0005eac6a1a251d8.journal
-rw-r-----.  1 root root   3762016 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000e7a-0005eac6a1a35a29.journal
-rw-r-----.  1 root root   3875720 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000e95-0005eac6a1a40b49.journal
-rw-r-----.  1 root root   3741248 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000f50-0005eac6a1b90a20.journal
-rw-r-----.  1 root root   3768264 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000f51-0005eac6a1b13145.journal
-rw-r-----.  1 root root   3867488 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000000f70-0005eac6a1b333bb.journal
-rw-r-----.  1 root root   4389920 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001009-0005eac6a1b9f417.journal
-rw-r-----.  1 root root   3760384 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000013fb-0005eac6a1ee8468.journal
-rw-r-----.  1 root root   3741008 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001412-0005eac6a1ee9299.journal
-rw-r-----.  1 root root   3740080 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001415-0005eac6a1f098c8.journal
-rw-r-----.  1 root root   3750160 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001417-0005eac6a1ee9319.journal
-rw-r-----.  1 root root   3739496 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001422-0005eac6a1f0ead0.journal
-rw-r-----.  1 root root   3761288 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001423-0005eac6a1ee9e53.journal
-rw-r-----.  1 root root   3739272 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000143e-0005eac6a1f17b58.journal
-rw-r-----.  1 root root   3740808 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000143f-0005eac6a1eed8f0.journal
-rw-r-----.  1 root root   3740216 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001442-0005eac6a1f18af8.journal
-rw-r-----.  1 root root   3740824 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001444-0005eac6a1eed93e.journal
-rw-r-----.  1 root root   3740216 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001447-0005eac6a1f1ae20.journal
-rw-r-----.  1 root root   3866048 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001449-0005eac6a1eed989.journal
-rw-r-----.  1 root root   3766888 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000014e3-0005eac6a1ef9937.journal
-rw-r-----.  1 root root   3904416 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001506-0005eac6a1efc818.journal
-rw-r-----.  1 root root   3997592 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000015ff-0005eac6a1f2aa5a.journal
-rw-r-----.  1 root root   3860392 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000175e-0005eac6a1f7f08f.journal
-rw-r-----.  1 root root   3778496 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000017d9-0005eac6a1fdba55.journal
-rw-r-----.  1 root root   3749000 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001809-0005eac6a1fdfcb8.journal
-rw-r-----.  1 root root   3740464 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001817-0005eac6a20ebba0.journal
-rw-r-----.  1 root root   3740408 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000181a-0005eac6a1fdff9b.journal
-rw-r-----.  1 root root   3838784 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000181d-0005eac6a1fe006a.journal
-rw-r-----.  1 root root   3746512 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000018af-0005eac6a1fe426f.journal
-rw-r-----.  1 root root   3746960 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000018b8-0005eac6a1fe45bc.journal
-rw-r-----.  1 root root   3744704 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000018c2-0005eac6a1fe4965.journal
-rw-r-----.  1 root root   3753848 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000018ca-0005eac6a1fe4ec4.journal
-rw-r-----.  1 root root   3777320 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000018dc-0005eac6a1fe56f6.journal
-rw-r-----.  1 root root   3791976 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001911-0005eac6a1ff31bb.journal
-rw-r-----.  1 root root   3785008 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001959-0005eac6a20086ca.journal
-rw-r-----.  1 root root   3742888 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001989-0005eac6a202ea52.journal
-rw-r-----.  1 root root   3742880 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000198c-0005eac6a202fcdc.journal
-rw-r-----.  1 root root   3742888 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-000000000000198f-0005eac6a203158d.journal
-rw-r-----.  1 root root   3769048 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001992-0005eac6a2033c6f.journal
-rw-r-----.  1 root root   3748616 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000019b7-0005eac6a2061567.journal
-rw-r-----.  1 root root   3741528 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000019c1-0005eac6a20688ca.journal
-rw-r-----.  1 root root   3742160 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000019c4-0005eac6a206b994.journal
-rw-r-----.  1 root root   3744216 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000019c8-0005eac6a2071551.journal
-rw-r-----.  1 root root   3786656 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-00000000000019cd-0005eac6a207406a.journal
-rw-r-----.  1 root root   3818856 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001a04-0005eac6a20b3547.journal
-rw-r-----.  1 root root   3745768 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001a66-0005eac6a2105939.journal
-rw-r-----.  1 root root   3823640 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001a6c-0005eac6a21076d4.journal
-rw-r-----.  1 root root   3804424 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001ad6-0005eac6a21327ac.journal
-rw-r-----.  1 root root   3743784 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001b26-0005eac6a216e6ff.journal
-rw-r-----.  1 root root   3744752 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001b29-0005eac6a216fac0.journal
-rw-r-----.  1 root root   3749896 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001b2d-0005eac6a2170441.journal
-rw-r-----.  1 root root   3782232 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001b36-0005eac6a217182d.journal
-rw-r-----.  1 root root   3765536 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001b64-0005eac6a217d34f.journal
-rw-r-----.  1 root root   3742240 Oct 11 18:46 test@2cdacd4e2f8848d39d6567c88d55d6c5-0000000000001b84-0005eac6a2247238.journal
-rw-r-----.  1 root root   8388608 Oct 11 18:46 test.journal
drwxr-xr-x. 14 root root      4096 Oct 11 18:40 unprivileged-nspawn-root
```

It gets so bad, that the test suite  on C8S fails half-way through, since there's over 30 GB of journals and the storage has only ~40 GB available for the whole system.

mrc0mmand commented:
```
# SYSTEMD_LOG_LEVEL=debug ~/systemd/build/systemd-journal-remote -o test.journal --getter "/root/systemd/build/journalctl -o export -D root/var/log/journal" |& tee log.txt
# grep smaller log.txt
Realtime timestamp 1665513672302000 smaller than previous realtime timestamp 1665513672687524, refusing entry.
Realtime timestamp 1665513672321000 smaller than previous realtime timestamp 1665513672705635, refusing entry.
Realtime timestamp 1665513672265491 smaller than previous realtime timestamp 1665513672698000, refusing entry.
Realtime timestamp 1665513675328000 smaller than previous realtime timestamp 1665513675333764, refusing entry.
Realtime timestamp 1665513678226000 smaller than previous realtime timestamp 1665513678231486, refusing entry.
Realtime timestamp 1665513678243000 smaller than previous realtime timestamp 1665513678245803, refusing entry.
Realtime timestamp 1665513680161000 smaller than previous realtime timestamp 1665513680164499, refusing entry.
Realtime timestamp 1665513684969000 smaller than previous realtime timestamp 1665513684970659, refusing entry.
Realtime timestamp 1665513685379000 smaller than previous realtime timestamp 1665513685385888, refusing entry.
Realtime timestamp 1665513686302000 smaller than previous realtime timestamp 1665513686304477, refusing entry.
Realtime timestamp 1665513687739000 smaller than previous realtime timestamp 1665513687743351, refusing entry.
Realtime timestamp 1665513690727000 smaller than previous realtime timestamp 1665513690729514, refusing entry.
Realtime timestamp 1665513690819000 smaller than previous realtime timestamp 1665513690821260, refusing entry.
Realtime timestamp 1665513690831000 smaller than previous realtime timestamp 1665513690832313, refusing entry.
Realtime timestamp 1665513690915961 smaller than previous realtime timestamp 1665513690916673, refusing entry.
Realtime timestamp 1665513690924841 smaller than previous realtime timestamp 1665513690926797, refusing entry.
Realtime timestamp 1665513690934000 smaller than previous realtime timestamp 1665513690936656, refusing entry.
Realtime timestamp 1665513690935819 smaller than previous realtime timestamp 1665513690958299, refusing entry.
Realtime timestamp 1665513691147000 smaller than previous realtime timestamp 1665513691153949, refusing entry.
Realtime timestamp 1665513693204000 smaller than previous realtime timestamp 1665513707896950, refusing entry.
Realtime timestamp 1665513696123000 smaller than previous realtime timestamp 1665513707920848, refusing entry.
Realtime timestamp 1665513696727000 smaller than previous realtime timestamp 1665513707933864, refusing entry.
Realtime timestamp 1665513706930000 smaller than previous realtime timestamp 1665513707945468, refusing entry.
Realtime timestamp 1665513707505000 smaller than previous realtime timestamp 1665513707958630, refusing entry.
Realtime timestamp 1665513695799168 smaller than previous realtime timestamp 1665513707918000, refusing entry.
Realtime timestamp 1665513695801410 smaller than previous realtime timestamp 1665513707975219, refusing entry.
Realtime timestamp 1665513695807751 smaller than previous realtime timestamp 1665513707989645, refusing entry.
Realtime timestamp 1665513695810246 smaller than previous realtime timestamp 1665513708001545, refusing entry.
Realtime timestamp 1665513695918574 smaller than previous realtime timestamp 1665513708052532, refusing entry.
Realtime timestamp 1665513695931856 smaller than previous realtime timestamp 1665513708185000, refusing entry.
Realtime timestamp 1665513695932715 smaller than previous realtime timestamp 1665513708959790, refusing entry.
Realtime timestamp 1665513695934770 smaller than previous realtime timestamp 1665513708997000, refusing entry.
Realtime timestamp 1665513695940274 smaller than previous realtime timestamp 1665513709061000, refusing entry.
Realtime timestamp 1665513695946788 smaller than previous realtime timestamp 1665513709147000, refusing entry.
Realtime timestamp 1665513695992800 smaller than previous realtime timestamp 1665513709473000, refusing entry.
Realtime timestamp 1665513709513317 smaller than previous realtime timestamp 1665513709919000, refusing entry.
Realtime timestamp 1665513709537676 smaller than previous realtime timestamp 1665513709949000, refusing entry.
Realtime timestamp 1665513709586819 smaller than previous realtime timestamp 1665513709997000, refusing entry.
Realtime timestamp 1665513710155996 smaller than previous realtime timestamp 1665513710193000, refusing entry.
Realtime timestamp 1665513710278257 smaller than previous realtime timestamp 1665513710278509, refusing entry.
Realtime timestamp 1665513710375474 smaller than previous realtime timestamp 1665513710377000, refusing entry.
Realtime timestamp 1665513710384294 smaller than previous realtime timestamp 1665513710386000, refusing entry.
Realtime timestamp 1665513712027000 smaller than previous realtime timestamp 1665513712030447, refusing entry.
Realtime timestamp 1665513712024828 smaller than previous realtime timestamp 1665513712027000, refusing entry.
Realtime timestamp 1665513718029806 smaller than previous realtime timestamp 1665513718037809, refusing entry.
Realtime timestamp 1665513718051000 smaller than previous realtime timestamp 1665513718056614, refusing entry.
Realtime timestamp 1665513718039435 smaller than previous realtime timestamp 1665513718052000, refusing entry.
Realtime timestamp 1665513718255000 smaller than previous realtime timestamp 1665513718259363, refusing entry.
Realtime timestamp 1665513718248692 smaller than previous realtime timestamp 1665513718255000, refusing entry.
Realtime timestamp 1665513718681299 smaller than previous realtime timestamp 1665513718689344, refusing entry.
Realtime timestamp 1665513718748000 smaller than previous realtime timestamp 1665513718751681, refusing entry.
Realtime timestamp 1665513718753000 smaller than previous realtime timestamp 1665513718757887, refusing entry.
Realtime timestamp 1665513718723928 smaller than previous realtime timestamp 1665513718755000, refusing entry.
Realtime timestamp 1665513718731809 smaller than previous realtime timestamp 1665513718797000, refusing entry.
Realtime timestamp 1665513718839814 smaller than previous realtime timestamp 1665513718985000, refusing entry.
Realtime timestamp 1665513718842366 smaller than previous realtime timestamp 1665513719068744, refusing entry.
Realtime timestamp 1665513718854364 smaller than previous realtime timestamp 1665513719265000, refusing entry.
Realtime timestamp 1665513720667000 smaller than previous realtime timestamp 1665513720838129, refusing entry.
Realtime timestamp 1665513719680748 smaller than previous realtime timestamp 1665513720766000, refusing entry.
Realtime timestamp 1665513719747032 smaller than previous realtime timestamp 1665513720889000, refusing entry.
Realtime timestamp 1665513719814697 smaller than previous realtime timestamp 1665513720926000, refusing entry.
Realtime timestamp 1665513719860041 smaller than previous realtime timestamp 1665513720965000, refusing entry.
Realtime timestamp 1665513721236000 smaller than previous realtime timestamp 1665513721236474, refusing entry.
Realtime timestamp 1665513720721733 smaller than previous realtime timestamp 1665513721236000, refusing entry.
Realtime timestamp 1665513720853435 smaller than previous realtime timestamp 1665513721290000, refusing entry.
Realtime timestamp 1665513721295895 smaller than previous realtime timestamp 1665513721500615, refusing entry.
Realtime timestamp 1665513724740712 smaller than previous realtime timestamp 1665513724844941, refusing entry.
Realtime timestamp 1665513724744345 smaller than previous realtime timestamp 1665513724877224, refusing entry.
Realtime timestamp 1665513724877000 smaller than previous realtime timestamp 1665513724882444, refusing entry.
Realtime timestamp 1665513724744473 smaller than previous realtime timestamp 1665513724880000, refusing entry.
Realtime timestamp 1665513724898000 smaller than previous realtime timestamp 1665513724900188, refusing entry.
Realtime timestamp 1665513724747347 smaller than previous realtime timestamp 1665513724898000, refusing entry.
Realtime timestamp 1665513724935000 smaller than previous realtime timestamp 1665513724938360, refusing entry.
Realtime timestamp 1665513724762352 smaller than previous realtime timestamp 1665513724935000, refusing entry.
Realtime timestamp 1665513724939000 smaller than previous realtime timestamp 1665513724945647, refusing entry.
Realtime timestamp 1665513724762430 smaller than previous realtime timestamp 1665513724942000, refusing entry.
Realtime timestamp 1665513724948000 smaller than previous realtime timestamp 1665513724953498, refusing entry.
Realtime timestamp 1665513724762505 smaller than previous realtime timestamp 1665513724951000, refusing entry.
Realtime timestamp 1665513724811575 smaller than previous realtime timestamp 1665513725151795, refusing entry.
Realtime timestamp 1665513724823576 smaller than previous realtime timestamp 1665513725201000, refusing entry.
Realtime timestamp 1665513725012570 smaller than previous realtime timestamp 1665513725524595, refusing entry.
Realtime timestamp 1665513725358223 smaller than previous realtime timestamp 1665513725987634, refusing entry.
Realtime timestamp 1665513725737557 smaller than previous realtime timestamp 1665513726572686, refusing entry.
Realtime timestamp 1665513725754552 smaller than previous realtime timestamp 1665513726808000, refusing entry.
Realtime timestamp 1665513726852000 smaller than previous realtime timestamp 1665513726864759, refusing entry.
Realtime timestamp 1665513725755291 smaller than previous realtime timestamp 1665513726852000, refusing entry.
Realtime timestamp 1665513725755498 smaller than previous realtime timestamp 1665513726876849, refusing entry.
Realtime timestamp 1665513725772399 smaller than previous realtime timestamp 1665513727101000, refusing entry.
Realtime timestamp 1665513725773244 smaller than previous realtime timestamp 1665513727114000, refusing entry.
Realtime timestamp 1665513725774181 smaller than previous realtime timestamp 1665513727130000, refusing entry.
Realtime timestamp 1665513725775556 smaller than previous realtime timestamp 1665513727145000, refusing entry.
Realtime timestamp 1665513725777654 smaller than previous realtime timestamp 1665513727186000, refusing entry.
Realtime timestamp 1665513725833659 smaller than previous realtime timestamp 1665513727266000, refusing entry.
Realtime timestamp 1665513725920970 smaller than previous realtime timestamp 1665513727373000, refusing entry.
Realtime timestamp 1665513726077522 smaller than previous realtime timestamp 1665513727451000, refusing entry.
Realtime timestamp 1665513726082268 smaller than previous realtime timestamp 1665513727457000, refusing entry.
Realtime timestamp 1665513726088589 smaller than previous realtime timestamp 1665513727463000, refusing entry.
Realtime timestamp 1665513726098543 smaller than previous realtime timestamp 1665513727469000, refusing entry.
Realtime timestamp 1665513726285159 smaller than previous realtime timestamp 1665513727528000, refusing entry.
Realtime timestamp 1665513726314698 smaller than previous realtime timestamp 1665513727543000, refusing entry.
Realtime timestamp 1665513726327188 smaller than previous realtime timestamp 1665513727547000, refusing entry.
Realtime timestamp 1665513726350673 smaller than previous realtime timestamp 1665513727565000, refusing entry.
Realtime timestamp 1665513726361706 smaller than previous realtime timestamp 1665513727585000, refusing entry.
Realtime timestamp 1665513726620999 smaller than previous realtime timestamp 1665513727661000, refusing entry.
Realtime timestamp 1665513726957881 smaller than previous realtime timestamp 1665513727791000, refusing entry.
Realtime timestamp 1665513726965460 smaller than previous realtime timestamp 1665513727800000, refusing entry.
Realtime timestamp 1665513727141804 smaller than previous realtime timestamp 1665513727971693, refusing entry.
Realtime timestamp 1665513727387391 smaller than previous realtime timestamp 1665513728101000, refusing entry.
Realtime timestamp 1665513727392448 smaller than previous realtime timestamp 1665513728107000, refusing entry.
Realtime timestamp 1665513727394881 smaller than previous realtime timestamp 1665513728112000, refusing entry.
Realtime timestamp 1665513727399981 smaller than previous realtime timestamp 1665513728125000, refusing entry.
Realtime timestamp 1665513727447887 smaller than previous realtime timestamp 1665513728184000, refusing entry.
Realtime timestamp 1665513728275000 smaller than previous realtime timestamp 1665513728647529, refusing entry.
Realtime timestamp 1665513727465519 smaller than previous realtime timestamp 1665513728293000, refusing entry.
```

This might explain the crapton of journals.

mrc0mmand commented:
Also, this is not limited to CI journals, it can be reproduced locally as well:
```
# ls -lh
total 0
# ~/systemd/build/systemd-journal-remote -o test.journal --getter "/root/systemd/build/journalctl -o export -b"
Spawning getter /root/systemd/build/journalctl -o export -b...
Finishing after writing 232409 entries
# ls -lh
total 4.9G
-rw-r-----. 1 root root 3.8M Oct 11 18:54 test@13d6cd25aaf94e8f847703be08858476-0000000000000001-0005eac4f5fa605d.journal
-rw-r-----. 1 root root 3.7M Oct 11 18:54 test@13d6cd25aaf94e8f847703be08858476-000000000000021c-0005eac4f5ff2148.journal
-rw-r-----. 1 root root 3.7M Oct 11 18:54 test@13d6cd25aaf94e8f847703be08858476-00000000000002a8-0005eac4f625a4a8.journal
...
-rw-r-----. 1 root root 3.6M Oct 11 18:56 test@13d6cd25aaf94e8f847703be08858476-0000000000038b77-0005eac67c9e4688.journal
-rw-r-----. 1 root root 3.6M Oct 11 18:56 test@13d6cd25aaf94e8f847703be08858476-0000000000038b7f-0005eac67c9e7568.journal
-rw-r-----. 1 root root 3.6M Oct 11 18:56 test@13d6cd25aaf94e8f847703be08858476-0000000000038b94-0005eac67ca02318.journal
-rw-r-----. 1 root root 3.6M Oct 11 18:56 test@13d6cd25aaf94e8f847703be08858476-0000000000038b97-0005eac67ca051f8.journal
-rw-r-----. 1 root root 3.6M Oct 11 18:56 test@13d6cd25aaf94e8f847703be08858476-0000000000038ba0-0005eac67e685440.journal
-rw-r-----. 1 root root 3.6M Oct 11 18:56 test@13d6cd25aaf94e8f847703be08858476-0000000000038ba5-0005eac67eb0bf50.journal
-rw-r-----. 1 root root 3.6M Oct 11 18:56 test@13d6cd25aaf94e8f847703be08858476-0000000000038ba9-0005eac67eb105a0.journal
-rw-r-----. 1 root root 8.0M Oct 11 18:56 test.journal
# ls | wc -l
1354
```

msizanoen1 commented:
> ```
> # SYSTEMD_LOG_LEVEL=debug ~/systemd/build/systemd-journal-remote -o test.journal --getter "/root/systemd/build/journalctl -o export -D root/var/log/journal" |& tee log.txt
> # grep smaller log.txt
> Realtime timestamp 1665513672302000 smaller than previous realtime timestamp 1665513672687524, refusing entry.
> Realtime timestamp 1665513672321000 smaller than previous realtime timestamp 1665513672705635, refusing entry.
> Realtime timestamp 1665513672265491 smaller than previous realtime timestamp 1665513672698000, refusing entry.
> Realtime timestamp 1665513675328000 smaller than previous realtime timestamp 1665513675333764, refusing entry.
> [snip]
> Realtime timestamp 1665513726327188 smaller than previous realtime timestamp 1665513727547000, refusing entry.
> Realtime timestamp 1665513726350673 smaller than previous realtime timestamp 1665513727565000, refusing entry.
> Realtime timestamp 1665513726361706 smaller than previous realtime timestamp 1665513727585000, refusing entry.
> Realtime timestamp 1665513726620999 smaller than previous realtime timestamp 1665513727661000, refusing entry.
> Realtime timestamp 1665513726957881 smaller than previous realtime timestamp 1665513727791000, refusing entry.
> Realtime timestamp 1665513726965460 smaller than previous realtime timestamp 1665513727800000, refusing entry.
> Realtime timestamp 1665513727141804 smaller than previous realtime timestamp 1665513727971693, refusing entry.
> Realtime timestamp 1665513727387391 smaller than previous realtime timestamp 1665513728101000, refusing entry.
> Realtime timestamp 1665513727392448 smaller than previous realtime timestamp 1665513728107000, refusing entry.
> Realtime timestamp 1665513727394881 smaller than previous realtime timestamp 1665513728112000, refusing entry.
> Realtime timestamp 1665513727399981 smaller than previous realtime timestamp 1665513728125000, refusing entry.
> Realtime timestamp 1665513727447887 smaller than previous realtime timestamp 1665513728184000, refusing entry.
> Realtime timestamp 1665513728275000 smaller than previous realtime timestamp 1665513728647529, refusing entry.
> Realtime timestamp 1665513727465519 smaller than previous realtime timestamp 1665513728293000, refusing entry.
> ```
> 
> This might explain the crapton of journals.

Can you send the journal files in question?

msizanoen1 commented:
It turns out systemd-journal-remote is mass rotating journal files because `journalctl -o export` is outputting bogus timestamps that's not guaranteed to be monotonically increasing. See last commit for details.

mrc0mmand commented:
It still seem to generate some extraneous journal(s):

Current main:
```
$ /home/mrc0mmand/repos/systemd/build-main/systemd-journal-remote -o test.journal --getter="/home/mrc0mmand/repos/systemd/build-main/journalctl -o export -b"
Spawning getter /home/mrc0mmand/repos/systemd/build-main/journalctl -o export -b...
Finishing after writing 8795 entries
$ ls -l
total 16388
-rw-r----- 1 mrc0mmand mrc0mmand 16777216 Oct 12 10:18 test.journal
```

This PR:
```
$ /home/mrc0mmand/repos/systemd/build/systemd-journal-remote -o test.journal --getter="/home/mrc0mmand/repos/systemd/build/journalctl -o export -b"
Spawning getter /home/mrc0mmand/repos/systemd/build/journalctl -o export -b...
Finishing after writing 8795 entries
total 13108
-rw-r----- 1 mrc0mmand mrc0mmand 5023328 Oct 12 10:19 test@a31bee0be923442cbfa6dc1ec5f116a2-0000000000000001-0005eab24368f919.journal
-rw-r----- 1 mrc0mmand mrc0mmand 8388608 Oct 12 10:19 test.journal
```

Debug log:
```
File /var/log/journal/1588e1d9d0b74acdbaada907b163b837/user-1000@0005e0fee50c7a7c-be5b8e6a2cf5741a.journal~ added.                                                                                                                             
File /var/log/journal/1588e1d9d0b74acdbaada907b163b837/user-1000@7dca8e1cb5474d19bf2f9e5ce3339b84-00000000002e1ed6-0005e092512d362e.journal added.                                                                                             
File /var/log/journal/1588e1d9d0b74acdbaada907b163b837/system@7d13ea5366ba46a7a8fc8b16728c07ee-000000000027e162-0005dda631583c34.journal added.                                                                                                
Journal filter: _BOOT_ID=119c2736ce9b4ea5a0567b44a17d87d0                                                                                                                                                                                      
Reserving 333 entries in field hash table.                                                                                                                                                                                                     
Reserving 233016 entries in data hash table.                                                                                                                                                                                                   
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=512B                                                                                                                                                   
Opened output file test.journal                                                                                                                                                                                                                
Creating source for fd:5 (test.journal)                                                                                                                                                                                                        
Watchdog is disabled.                                                                                                                                                                                                                          
systemd-journal-remote running as pid 360027                                                                                                                                                                                                   
Realtime timestamp 1665426233421878 smaller than previous realtime timestamp 1665426266021849, refusing entry.                                                                                                                                 
test.journal: Write failed, rotating: Text file busy                                                                                                                                                                                           
Reserving 333 entries in field hash table.                                                                                                                                                                                                     
Reserving 233016 entries in data hash table.                                                                                                                                                                                                   
test.journal: Successfully rotated journal                                                                                                                                                                                                     
Retrying write.                                                                                                                                                                                                                                
Compressed data object 1524 -> 880 using ZSTD                                                                                                                                                                                                  
Compressed data object 1380 -> 769 using ZSTD                                                                                                                                                                                                  
Compressed data object 1522 -> 882 using ZSTD                        
```

That's on my local machine, I'll try some VM as well.

msizanoen1 commented:
@mrc0mmand This is expect behavior. The CI scripts use the `systemd.clock_usec` parameter which may cause the same effect as a clock rollback during initrd->system transition: https://github.com/systemd/systemd-centos-ci/blob/dd8936b857ba653baecdf2d6cb79195339c7dc69/agent/bootstrap.sh#L261

msizanoen1 commented:
Can you retry it with a single input journal file, e.g. `journalctl --file /var/log/journal/<machine id>/system.journal`?

mrc0mmand commented:
> Can you retry it with a single input journal file, e.g. `<builddir>/journalctl --file /var/log/journal/<machine id>/system.journal`?

```
$ SYSTEMD_LOG_LEVEL=debug /home/mrc0mmand/repos/systemd/build/systemd-journal-remote -o test.journal --getter="/home/mrc0mmand/repos/systemd/build/journalctl -o export --file /var/log/journal/$(</etc/machine-id)/system.journal" |& tee output.log
Full config: SplitMode=none Key=n/a Cert=n/a Trust=n/a
Received 0 descriptors
Spawning getter /home/mrc0mmand/repos/systemd/build/journalctl -o export --file /var/log/journal/1588e1d9d0b74acdbaada907b163b837/system.journal...
Successfully forked off '(remote)' as PID 361599.
Fixed min_use=16.0M max_use=4.0G max_size=128.0M min_size=512.0K keep_free=4.0G n_max_files=100
Skipping PR_SET_MM, as we don't have privileges.
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=8B
File /var/log/journal/1588e1d9d0b74acdbaada907b163b837/system.journal added.
Journal filter: none
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=512B
Opened output file test.journal
Creating source for fd:5 (test.journal)
Watchdog is disabled.
systemd-journal-remote running as pid 361596
Realtime timestamp 1665426233421878 smaller than previous realtime timestamp 1665562644254781, refusing entry.
test.journal: Write failed, rotating: Text file busy
Reserving 333 entries in field hash table.
Reserving 233016 entries in data hash table.
test.journal: Successfully rotated journal
Retrying write.
mmap cache statistics: 242735 context cache hit, 3 window list hit, 1 miss
EOF reached with source test.journal (fd=5)
Closing test.journal (fd=5)
Writer ref count 2
Failed to remove source test.journal (type io) from epoll, ignoring: Bad file descriptor
0 active sources remaining
Finishing after writing 2715 entries
Closing journal file test.journal.
```

msizanoen1 commented:
Can you run a timestamp consistency check using `<builddir>/journalctl -o json-seq --file <file> | jq -r --seq '.__REALTIME_TIMESTAMP' | sort -n --check`?

mrc0mmand commented:
> journalctl -o json-seq --file <file> | jq -r --seq '.__REALTIME_TIMESTAMP' | sort -n --check

```
$ ../build/journalctl -o json-seq --file /var/log/journal/$(</etc/machine-id)/system.journal | jq -r --seq '.__REALTIME_TIMESTAMP' | wc -l
2715
$ ../build/journalctl -o json-seq --file /var/log/journal/$(</etc/machine-id)/system.journal | jq -r --seq '.__REALTIME_TIMESTAMP' | sort -n --check
$ echo $?
0
```

msizanoen1 commented:
Can you create a [gist](https://gist.github.com) with the `system.journal` file as-is and provide the link here?

mrc0mmand commented:
> @mrc0mmand If you can't upload the system.journal file yet, are you sure that you deleted the test.journal file before each command? The default behavior for the journal file library is to continue with the existing file, not start from scratch.

I can't really share the journal without doing some cleanup, but in my case the issue appears to be caused by NTP:

```
$ journalctl --file /var/log/journal/1588e1d9d0b74acdbaada907b163b837/system.journal --no-hostname -o short-monotonic
[   29.133498] chronyd[1365]: System clock wrong by -32.600208 seconds
[   29.125457] systemd-journald[994]: Time jumped backwards, rotating.
[   29.269699] systemd-resolved[1262]: Clock change detected. Flushing caches.
[   29.269900] chronyd[1365]: System clock was stepped by -32.600208 seconds
[   29.269951] chronyd[1365]: System clock TAI offset set to 37 seconds
[   29.269983] rsyslogd[1314]: imjournal: journal files changed, reloading...  [v8.2204.0-2.fc36 try https://www.rsyslog.com/e/0 ]
[   29.270146] rsyslogd[1314]: imjournal: journal files changed, reloading...  [v8.2204.0-2.fc36 try https://www.rsyslog.com/e/0 ]
[   29.270188] rsyslogd[1314]: imjournal: journal files changed, reloading...  [v8.2204.0-2.fc36 try https://www.rsyslog.com/e/0 ]
[   30.765244] chronyd[1365]: Selected source 213.192.54.227 (2.fedora.pool.ntp.org)
...
```
```
$ journalctl --file /var/log/journal/1588e1d9d0b74acdbaada907b163b837/system.journal --no-hostname -o short
Oct 10 20:24:26 chronyd[1365]: System clock wrong by -32.600208 seconds
Oct 10 20:23:53 systemd-journald[994]: Time jumped backwards, rotating.
Oct 10 20:23:53 systemd-resolved[1262]: Clock change detected. Flushing caches.
Oct 10 20:23:53 chronyd[1365]: System clock was stepped by -32.600208 seconds
Oct 10 20:23:53 chronyd[1365]: System clock TAI offset set to 37 seconds
Oct 10 20:23:53 rsyslogd[1314]: imjournal: journal files changed, reloading...  [v8.2204.0-2.fc36 try https://www.rsyslog.com/e/0 ]
Oct 10 20:23:53 rsyslogd[1314]: imjournal: journal files changed, reloading...  [v8.2204.0-2.fc36 try https://www.rsyslog.com/e/0 ]
Oct 10 20:23:53 rsyslogd[1314]: imjournal: journal files changed, reloading...  [v8.2204.0-2.fc36 try https://www.rsyslog.com/e/0 ]
Oct 10 20:23:55 chronyd[1365]: Selected source 213.192.54.227 (2.fedora.pool.ntp.org)
...
```

Which I guess is expected-ish.

Anyway, if/when this lands, the artifact handling in various CIs will need to be tweaked, as now there's a high probability the exported journal will get fragmented across multiple files even with `systemd-journal-remote`'s coalescing (especially with tests that mess with the time, like TEST-30, TEST-45, TEST-53, etc.), /cc @keszybz, @bluca.

msizanoen1 commented:
Seems like the problem with your testing is indeed forgetting to delete the `test.journal` file:
```
 redacted@redacted  redacted   redacted  SYSTEMD_LOG_LEVEL=debug ./_build/systemd-journal-remote -o test.journal --getter "./_build/journalctl -o export --file /var/log/journal/$(</etc/machine-id)/system.journal"
Failed to open configuration file '/etc/systemd/journal-remote.conf': No such file or directory
Full config: SplitMode=none Key=n/a Cert=n/a Trust=n/a
Received 0 descriptors
Spawning getter ./_build/journalctl -o export --file /var/log/journal/redacted/system.journal...
Successfully forked off '(remote)' as PID 146680.
Skipping PR_SET_MM, as we don't have privileges.
Creating journal file test.journal on a btrfs file system, and copy-on-write is enabled. This is likely to slow down journal access substantially, please consider turning off the copy-on-write file attribute on the journal directory, using chattr +C.
Fixed min_use=16.0M max_use=4.0G max_size=128.0M min_size=512.0K keep_free=4.0G n_max_files=100
Reserving 333 entries in field hash table.
Journal effective settings seal=yes keyed_hash=yes compress=yes compress_threshold_bytes=8B
File /var/log/journal/redacted/system.journal added.
Journal filter: none
Reserving 233016 entries in data hash table.
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=512B
Opened output file test.journal
Creating source for fd:5 (test.journal)
Watchdog is disabled.
systemd-journal-remote running as pid 146679
Compressed data object 520 -> 256 using ZSTD
Compressed data object 606 -> 273 using ZSTD
Compressed data object 8393 -> 794 using ZSTD
Compressed data object 4111 -> 437 using ZSTD
Compressed data object 4226 -> 764 using ZSTD
Compressed data object 2023 -> 417 using ZSTD
mmap cache statistics: 412443 context cache hit, 3 window list hit, 1 miss
EOF reached with source test.journal (fd=5)
Closing test.journal (fd=5)
Writer ref count 2
Failed to remove source test.journal (type io) from epoll, ignoring: Bad file descriptor
0 active sources remaining
Finishing after writing 3864 entries
Closing journal file test.journal.
 redacted@redacted  redacted   redacted  SYSTEMD_LOG_LEVEL=debug ./_build/systemd-journal-remote -o test.journal --getter "./_build/journalctl -o export --file /var/log/journal/$(</etc/machine-id)/system.journal"
Failed to open configuration file '/etc/systemd/journal-remote.conf': No such file or directory
Full config: SplitMode=none Key=n/a Cert=n/a Trust=n/a
Received 0 descriptors
Spawning getter ./_build/journalctl -o export --file /var/log/journal/redacted/system.journal...
Successfully forked off '(remote)' as PID 146732.
Skipping PR_SET_MM, as we don't have privileges.
Fixed min_use=16.0M max_use=4.0G max_size=128.0M min_size=512.0K keep_free=4.0G n_max_files=100
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=512B
Opened output file test.journal
Creating source for fd:5 (test.journal)
Watchdog is disabled.
systemd-journal-remote running as pid 146731
Journal effective settings seal=yes keyed_hash=yes compress=yes compress_threshold_bytes=8B
File /var/log/journal/redacted/system.journal added.
Journal filter: none
Realtime timestamp 1665499819599180 smaller than previous realtime timestamp 1665587127276377, refusing entry.
test.journal: Write failed, rotating: Text file busy
Creating journal file test.journal on a btrfs file system, and copy-on-write is enabled. This is likely to slow down journal access substantially, please consider turning off the copy-on-write file attribute on the journal directory, using chattr +C.
Reserving 333 entries in field hash table.
Reserving 233016 entries in data hash table.
test.journal: Successfully rotated journal
Retrying write.
Compressed data object 520 -> 256 using ZSTD
Compressed data object 606 -> 273 using ZSTD
Compressed data object 8393 -> 794 using ZSTD
Compressed data object 4111 -> 437 using ZSTD
Compressed data object 4226 -> 764 using ZSTD
Compressed data object 2023 -> 417 using ZSTD
mmap cache statistics: 412443 context cache hit, 3 window list hit, 1 miss
EOF reached with source test.journal (fd=5)
Closing test.journal (fd=5)
Writer ref count 2
Failed to remove source test.journal (type io) from epoll, ignoring: Bad file descriptor
0 active sources remaining
Finishing after writing 3864 entries
Closing journal file test.journal.
 redacted@redacted  redacted   redacted  ls *.journal
test@redacted-0000000000000001-0005eac3651ee54c.journal  test.journal
 redacted@redacted  redacted   redacted  rm *.journal
 redacted@redacted  redacted   redacted  SYSTEMD_LOG_LEVEL=debug ./_build/systemd-journal-remote -o test.journal --getter "./_build/journalctl -o export --file /var/log/journal/$(</etc/machine-id)/system.journal"
Failed to open configuration file '/etc/systemd/journal-remote.conf': No such file or directory
Full config: SplitMode=none Key=n/a Cert=n/a Trust=n/a
Received 0 descriptors
Spawning getter ./_build/journalctl -o export --file /var/log/journal/redacted/system.journal...
Successfully forked off '(remote)' as PID 146915.
Creating journal file test.journal on a btrfs file system, and copy-on-write is enabled. This is likely to slow down journal access substantially, please consider turning off the copy-on-write file attribute on the journal directory, using chattr +C.
Skipping PR_SET_MM, as we don't have privileges.
Fixed min_use=16.0M max_use=4.0G max_size=128.0M min_size=512.0K keep_free=4.0G n_max_files=100
Reserving 333 entries in field hash table.
Reserving 233016 entries in data hash table.
Journal effective settings seal=yes keyed_hash=yes compress=yes compress_threshold_bytes=8B
File /var/log/journal/redacted/system.journal added.
Journal filter: none
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=512B
Opened output file test.journal
Creating source for fd:5 (test.journal)
Watchdog is disabled.
systemd-journal-remote running as pid 146914
Compressed data object 520 -> 256 using ZSTD
Compressed data object 606 -> 273 using ZSTD
Compressed data object 8393 -> 794 using ZSTD
Compressed data object 4111 -> 437 using ZSTD
Compressed data object 4226 -> 764 using ZSTD
Compressed data object 2023 -> 417 using ZSTD
mmap cache statistics: 412443 context cache hit, 3 window list hit, 1 miss
EOF reached with source test.journal (fd=5)
Closing test.journal (fd=5)
Writer ref count 2
Failed to remove source test.journal (type io) from epoll, ignoring: Bad file descriptor
0 active sources remaining
Finishing after writing 3864 entries
Closing journal file test.journal.
 redacted@redacted  redacted   redacted  ls *.journal
test.journal
 redacted@redacted  redacted   redacted 
```

msizanoen1 commented:
The fix for journal exporting has been spun off info #24996 but the commit remains here for CI purposes.

msizanoen1 commented:
@yuwata, @mrc0mmand: There hasn't been any response from a maintainer on this PR for a week. Are there still any interests in merging this?

yuwata commented:
Sorry for late response, but please wait. To make this not forgotten, I set a milestone for this PR.

poettering commented on src/libsystemd/sd-journal/journal-file.c:
```diff
@@ -2091,6 +2091,12 @@ static int journal_file_append_entry_internal(
+                                       "Realtime timestamp %" PRIu64 " smaller than previous realtime "
+                                       "timestamp %" PRIu64 ", refusing entry.",
+                                       ts->realtime, le64toh(f->header->tail_entry_realtime));
+
```
makes sense to add, but can you make up a new error code here? we already use ETXTBSY for journal files "from the future", but this is a bit different, it's internally non-consistent journal files.

Maybe EREMCHG? That's currently not used by the journal sources.

poettering replied:
note that this is half redundant, given write_to_journal() already does a similar check. would be good adding a comment there, to mention this.

(I think it makes sense to keep both logics in place, even if one is redundant, since they have a different focus: one looks at all log entries, the other one only to those hitting a specific file)

poettering commented on src/journal/journald-server.c:
```diff
@@ -827,8 +827,8 @@ static bool shall_try_append_again(JournalFile *f, int r) {
-        case -ETXTBSY:         /* Journal file is from the future */
-                log_ratelimit_warning(JOURNALD_LOG_RATELIMIT, "%s: Journal file is from the future, rotating.", f->path);
+        case -ETXTBSY:         /* Time jumped backwards relative to last journal entry */
+                log_ratelimit_warning(JOURNALD_LOG_RATELIMIT, "%s: Time jumped backwards relative to last journal entry, rotating.", f->path);
```
i'd really use a different error code, and leave ETXTBSY thing for the "future file"

poettering reviewed (changes requested):
looks good to me, but see comments

poettering commented:
(and sorry for the late review, there's so much stuff to review)

poettering commented:
managed_journal_file_open_reliably() also needs an update for the new error code

poettering commented:
excellent! thanks! lgtm.
